### ISSUE TITLE
How to decode recursive data structures?

### ISSUE THREAD

**Issue by svdHero** [nvie/decoders#510]
Hi there,

how would I decode recursive data structures, such as
```typescript
type Folder = {
    name: string,
    subfolders: Folder[]
}
```

Elm has the `lazy` decoder for this:
https://package.elm-lang.org/packages/elm/json/latest/Json-Decode#lazy

How would I solve this using this `decoders` package?

**Comment by nvie** [comment]
Great point! This has been on my list to implement, but I seem to have forgotten about it. I've implemented it in #511 and am still running a few tests with it. I'll make a new release with this lazy decoder included as soon as I'm happy with the results.

**Comment by nvie** [comment]
With this new decoder, the following should be possible:

```js
import { array, Decoder, guard, lazy, object, string } from "decoders";

type Folder = {
  name: string;
  subfolders: Folder[];
};

const folder: Decoder<Folder> = object({
  name: string,
  subfolders: array(lazy(() => folder)),
});
```

**Comment by nvie** [comment]
Now available in [v1.21.0](https://github.com/nvie/decoders/releases/tag/v1.21.0).

**Comment by svdHero** [comment]
Awesome. Thanks a lot.
